src/shred_network/shred_receiver.zig

```diff
@@ -41,68 +42,102 @@ pub const ShredReceiver = struct {
+            signal.event.set();
+        }
+
+        fn wait(signal: *ReceiverSignal, exit: ExitCondition) error{Exit}!void {
+            while (true) {
+                signal.event.timedWait(1 * std.time.ns_per_s) catch {};
+                if (exit.shouldExit()) return error.Exit;
+                if (signal.event.isSet()) return signal.event.reset();
+            }
+        }
+    };
+
     /// Run threads to listen/send over socket and handle all incoming packets.
     /// Returns when exit is set to true.
     pub fn run(self: *Self) !void {
         defer self.logger.err().log("exiting shred receiver");
         errdefer self.logger.err().log("error in shred receiver");
 
-        var response_sender = try SocketThread
-            .initSender(self.allocator, self.logger.unscoped(), self.repair_socket, self.exit);
-        defer response_sender.deinit(self.allocator);
-        var repair_receiver = try SocketThread
-            .initReceiver(self.allocator, self.logger.unscoped(), self.repair_socket, self.exit);
-        defer repair_receiver.deinit(self.allocator);
+        var receive_signal = ReceiverSignal{};
+        const exit = ExitCondition{ .unordered = self.exit };
+
+        // Cretae pipe from response_sender -> repair_socket
+        const response_sender = try Channel(Packet).create(self.allocator);
+        defer response_sender.destroy();
+
+        const response_sender_pipe = try SocketPipe.init(
+            self.allocator,
+            .sender,
+            self.logger.unscoped(),
+            self.repair_socket,
+            response_sender,
+            exit,
+        );
+        defer response_sender_pipe.deinit(self.allocator);
+
+        // Create pipe from repair_socket -> response_receiver.
+        const response_receiver = try Channel(Packet).create(self.allocator);
+        response_receiver.send_hook = &receive_signal.hook;
+        defer response_receiver.destroy();
 
-        var turbine_receivers: [NUM_TVU_RECEIVERS]SocketThread = undefined;
-        for (0..NUM_TVU_RECEIVERS) |i| {
-            turbine_receivers[i] = try SocketThread.initReceiver(
+        const response_receiver_pipe = try SocketPipe.init(
+            self.allocator,
+            .receiver,
+            self.logger.unscoped(),
+            self.repair_socket,
+            response_receiver,
+            exit,
+        );
+        defer response_receiver_pipe.deinit(self.allocator);
+
+        // Create N pipes from turbine_socket -> turbine_receiver.
+        const turbine_receiver = try Channel(Packet).create(self.allocator);
+        turbine_receiver.send_hook = &receive_signal.hook;
+        defer turbine_receiver.destroy();
+
+        var turbine_receiver_pipes: std.BoundedArray(*SocketPipe, NUM_TVU_RECEIVERS) = .{};
+        defer for (turbine_receiver_pipes.slice()) |pipe| pipe.deinit(self.allocator);
+        for (0..NUM_TVU_RECEIVERS) |_| {
```

Comment:
Given that we're moving towards reducing the number of threads dedicated to network IO, I'm curious on your opinion about using multiple threads here.

Agave actually uses eight threads for this. I decided to support the same logic of being able to run on multiple threads, basically just for feature parity, but I kept it a bit more conservative at 2 threads since 8 felt like overkill. Now I'm thinking we could just simplify this and just move it to one thread.

I don't know why they dedicated multiple threads to this. Turbine does receive a lot of data. But occasionally repair is expected to receive a lot more data than turbine ever receives. So I don't get it. Is it conceivable that a single CPU would be insufficient for simply reading packets and writing them to a channel?

Reply:
Maybe to offset inter-thread/channel communication? Not sure. I also think it could move to one thread. The saturation would matter more if the recvfrom CPU was doing many thread unparks on the channel's ResetEvent. Or maybe if its doing other IO work.

Reply:
https://github.com/Syndica/sig/pull/486/commits/00592b9394273a16c0384109fb7295acd8158df4